### PR TITLE
Add support for whitelisting origins

### PR DESCRIPTION
**Pull request by agrubi**
The option can be used to whitelist href origins when start() or perform() is called on the client.
Add unit tests

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2,34 +2,64 @@ const Entity = require('siren-parser');
...
+   * @param method
+   * @param [fieldValues]
+   * @returns {Promise<Entity>}
+   */
   async perform(href, method, fieldValues) {
+    verifyOrigin(href, this._options.originWhitelist);
+    return this._perform(href, method, fieldValues);
+  }
+
+  async _perform(href, method, fieldValues) {
```
Comment (agrubi) [inline index.js]: is this ok as is or should it be made completely private (a method outside the class)?

Comment (agrubi) [reply index.js]: another similar question: should we make attachClient() below private or prefix it with underscore? Do we expect it to be called by the users of the class?

Comment (ThomasMorrison) [reply index.js]: I think attachClient should stay public, as it does support a use case where you want to attach an existing entity to this Client instance. 

I think `_perform` could be fully private however, as we really shouldnt be circumventing this sort of thing. If you need to circumvent this, make a new client with no origin whitelist.

Comment (agrubi) [reply index.js]: sounds good to me. Will work on making it private.

File: index.js [hunk 2]
```diff
@@ -2,45 +2,66 @@ const Entity = require('siren-parser');
...
+   * @property body response body
+   * @property contentType Content-Type response header
+   *
+   * @callback requestFn
+   * @param {string} href (required) href of the resource
    * @param {string} method (required) HTTP request verb ('GET', 'POST', 'PUT', 'PATCH', 'DELETE')
-   * @param {object} fieldValues (optional) field names and values for the request
-   * @return {object} body: response body, contentType: Content-Type response header
+   * @param {Object} [fieldValues] (optional) field names and values for the request
+   * @return {Response} body: response body, contentType: Content-Type response header
```
Comment (nathanhoel) [inline index.js]: you changed a bunch of the documentation here. Can you explain the changes?
The format of this documentation is defined in jsdocs.
I am not an expert on jsdocs but I thought the thing in the curly braces had to be a type.

File: index.js [hunk 3]
```diff
@@ -2,45 +2,66 @@ const Entity = require('siren-parser');
...
-  constructor(requestFn) {
+  constructor(requestFn, options = {}) {
     this._requestFn = requestFn;
+    this._options = options;
   }
 
+  /**
+   * Starts a client by fetching the provided href
+   * @param {string} href
+   * @returns {Promise<Entity>}
```
Comment (nathanhoel) [inline index.js]: should give a description of the return and a better description for the href

File: index.js [hunk 4]
```diff
@@ -2,45 +2,66 @@ const Entity = require('siren-parser');
...
-    const { body, contentType } = await this._requestFn(href, method, fieldValues);
-    if (!body || !contentType || !_isSiren(contentType)) {
-      return body;
-    }
-
-    const entity = new Entity(body);
-    this.attachClient(entity);
-
-    return entity;
+    _verifyOrigin(href, this._options.originWhitelist);
```
Comment (nathanhoel) [inline index.js]: I don't think this should verify the origin on every call - JUST the start call and let it perform or follow any link after that.

File: index.js (hunk 2 above)
Comment (agrubi) [reply index.js]: I was trying to find what are predefined types in JSDoc, and I found this medium post that mentions that suggests to use capitalized names for objects, and only use lower case for boolean, string, and number. I guess also for functions and callbacks. https://medium.com/@trukrs/type-safe-javascript-with-jsdoc-7a2a63209b76

So, in this case type of field values is an object, and type of body is Response, which is defined above. 

Square brackets are to mark values as optional.

File: index.js (hunk 4 above)
Comment (agrubi) [reply index.js]: yes, that's exactly what is happening

Comment (agrubi) [reply index.js]: only start and the first perform will do the verification

**Pull request merged**